jitbuilder/lljb/src/IRVisitor.cpp

```diff
@@ -0,0 +1,559 @@
+   if (_methodBuilder->isIndirectLoadOrStore(dest))
+      {
+      _builder->StoreAt(
+               _builder->IndexAt(
+                  _td->PointerTo(_methodBuilder->getIlType(_td,value->getType())),
+                  _methodBuilder->getIlValue(dest),
+                  _builder->ConstInt32(0)),
+               ilValue);
+      }
+   else
+      {
+      _builder->Store(_methodBuilder->getLocalNameFromValue(dest), ilValue);
+      }
+   }
+
+void
+IRVisitor::visitBinaryOperator(llvm::BinaryOperator &I)
+   {
+   TR::IlValue * lhs = getIlValue(I.getOperand(0));
+   TR::IlValue * rhs = getIlValue(I.getOperand(1));
+   TR::IlValue * result = nullptr;
+   switch (I.getOpcode())
+      {
+      case llvm::Instruction::BinaryOps::Add:
+      case llvm::Instruction::BinaryOps::FAdd:
+         result = _builder->Add(lhs,rhs);
+         break;
+      case llvm::Instruction::BinaryOps::Sub:
+      case llvm::Instruction::BinaryOps::FSub:
+         result = _builder->Sub(lhs, rhs);
+         break;
+      case llvm::Instruction::BinaryOps::Mul:
+      case llvm::Instruction::BinaryOps::FMul:
+         result = _builder->Mul(lhs, rhs);
+         break;
+      case llvm::Instruction::BinaryOps::SDiv:
+      case llvm::Instruction::BinaryOps::FDiv:
+      case llvm::Instruction::BinaryOps::UDiv:
+         result = _builder->Div(lhs, rhs);
+         break;
+      case llvm::Instruction::BinaryOps::Shl:
+         if (!I.getOperand(1)->getType()->isIntegerTy(32))
+               rhs = _builder->ConvertTo(_td->Int32, rhs);
+         result = _builder->ShiftL(lhs, rhs);
+         break;
+      case llvm::Instruction::BinaryOps::AShr:
+         if (!I.getOperand(1)->getType()->isIntegerTy(32))
+               rhs = _builder->ConvertTo(_td->Int32, rhs);
+         result = _builder->ShiftR(lhs, rhs);
+         break;
+      case llvm::Instruction::BinaryOps::LShr:
+         if (!I.getOperand(1)->getType()->isIntegerTy(32))
+               rhs = _builder->ConvertTo(_td->Int32, rhs);
+         result = _builder->UnsignedShiftR(lhs, rhs);
+         break;
+      case llvm::Instruction::BinaryOps::And:
+         result = _builder->And(lhs, rhs);
+         break;
+      case llvm::Instruction::BinaryOps::Or:
+         result = _builder->Or(lhs, rhs);
+         break;
+      case llvm::Instruction::BinaryOps::Xor:
+         result = _builder->Xor(lhs, rhs);
+         break;
+      default:
+         llvm::outs() << "Instruction being visited: " << I << "\n";
+         assert(0 && "Unknown binary operand");
+         break;
+      }
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitCmpInst(llvm::CmpInst &I)
+   {
+   TR::IlValue * lhs = getIlValue(I.getOperand(0));
+   TR::IlValue * rhs = getIlValue(I.getOperand(1));
+
+   TR::IlValue * result = nullptr;
+
+   switch (I.getPredicate())
+      {
+      case llvm::CmpInst::Predicate::ICMP_EQ:
+      case llvm::CmpInst::Predicate::FCMP_OEQ:
+      case llvm::CmpInst::Predicate::FCMP_UEQ:
```

Comment:
We should really create Unordered comparison operations in JitBuilder that leverage the unordered fcmp opcodes in the OMR compiler rather than map both ordered and unordered fp compares to the same JitBuilder comparisons. Any code that's senstiive to unordered floating point values is likely to behave incorrectly here (at least if a NaN shows up), so maybe it's better to assert on those opcodes until the appropriate operations are added to JitBuilder?

Reply:
Done. I have also changed one of the test cases that used to generate `FCMP_UEQ` LLVM instruction.